File: LeetCode/Linked List Cicle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        hare = tortoise = head
        
        while hare and hare.next:
            tortoise = tortoise.next
            hare = hare.next.next
            
            if (hare == tortoise):
                break
            
        if (not hare or not hare.next):
            return None
        
        tortoise = head
        
        while (hare != tortoise):
            hare = hare.next
            tortoise = tortoise.next
        
        return tortoise
    
        

File: LeetCode/Linked List Cicle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *current = head;
        map<ListNode*, bool> visited;
        
        while(current && !visited[current]) {
            visited[current] = true;
            current = current->next;
        }
        
        return current;
    }
};